### PR TITLE
fix: es5

### LINKED ISSUE #12885: arrowFunction exsit in es5 mode？
# Bug report

> webpack.config.js
`{
    target: 'es5'
}`

> build result
```
1. __webpack_require__.a(module, async (__webpack_handle_async_dependencies__) => {}
   2. var wrapDeps = function(deps) { return deps.map(function(dep) {
	if(dep !== null && typeof dep === "object") {
		if(dep[webpackThen]) return dep;
		     if(dep.then) {
      		         var queue = [], result;
					dep.then(function(r) {
						obj[webpackExports] = r;
						completeQueue(queue);
 						queue = 0;
 					});
					var obj = { [webpackThen]: (fn, reject) => { queueFunction(queue, fn); dep.catch(reject); } };
 					return obj;
				}
 			}
			return { [webpackThen]: (fn) => { completeFunction(fn); }, [webpackExports]: dep };
 		}); };
```

> files at:
1. webpack/lib/dependencies/HarmonyCompatibilityDependency.js
2. webpack/lib/runtime/AsyncModuleRuntimeModule.js

> why not run supportsArrowFunction()?
1. runtimeTemplate.supportsArrowFunction()
2. Can traditional functions cause other errors ？

```
new InitFragment(
	`${RuntimeGlobals.asyncModule}(${module.moduleArgument}, **async (__webpack_handle_async_dependencies__) => {\n`**,
	InitFragment.STAGE_ASYNC_BOUNDARY,
	0,
	undefined,
	module.buildMeta.async
		? `\n__webpack_handle_async_dependencies__();\n}, 1);`
		: "\n});"
)
```

```
`var wrapDeps = ${runtimeTemplate.returningFunction(
      `deps.map(${runtimeTemplate.basicFunction("dep", [
	      'if(dep !== null && typeof dep === "object") {',
	      Template.indent([
		      "if(dep[webpackThen]) return dep;",
		      "if(dep.then) {",
		      Template.indent([
			      "var queue = [], result;",
			      `dep.then(${runtimeTemplate.basicFunction("r", [
				      "obj[webpackExports] = r;",
				      "completeQueue(queue);",
				      "queue = 0;"
			      ])});`,
			      "**var obj = { [webpackThen]: (fn, reject) => { queueFunction(queue, fn); dep.catch(reject); } };**",
			      "return obj;"
		      ]),
		      "}"
	      ]),
	      "}",
	      "**return { [webpackThen]: (fn) => { completeFunction(fn); }, [webpackExports]: dep };**"
      ])})`,
      "deps"
)};`,
```

**Other relevant information:**
webpack version:  5.25.0
Node.js version: 12
Operating System: mac
Additional tools:

english is poor, sry

### PR DESCRIPTION
**Pull request by alexander-akait**
**What kind of change does this PR introduce?**

bugfix, but we still have `async` here, but not related to this fix, fixes #12885

**Did you add tests for your changes?**

hard to test, I think we can use `acorn` to check syntax (try to parse based on `es` target), if we will face with it again I will write couple tests

**Does this PR introduce a breaking change?**

No

**What needs to be documented once your changes are merged?**

No

### CODE REVIEW

**Review by sokra — CHANGES_REQUESTED** [review]
Use the runtimeTemplate.basicFunction or runtimeTemplate.returningFunction helpers instead of selecting based on arrow function. That makes it more readable and less prone to errors in the duplicated code

File: lib/runtime/AsyncModuleRuntimeModule.js [hunk 1]
```diff
@@ -24,49 +24,49 @@ class AsyncModuleRuntimeModule extends HelperRuntimeModule {
...
+				"queue ? queue.push(fn) : completeFunction(fn)"
+			)};`,
 			`var wrapDeps = ${runtimeTemplate.returningFunction(
 				`deps.map(${runtimeTemplate.basicFunction("dep", [
 					'if(dep !== null && typeof dep === "object") {',
 					Template.indent([
 						"if(dep[webpackThen]) return dep;",
 						"if(dep.then) {",
 						Template.indent([
-							"var queue = [], result;",
```
Comment (alexander-akait) [inline lib/runtime/AsyncModuleRuntimeModule.js:57]: `result` is not used anywhere, so I just remove it

File: lib/runtime/CreateFakeNamespaceObjectRuntimeModule.js [hunk 2]
```diff
@@ -51,14 +48,13 @@ class CreateFakeNamespaceObjectRuntimeModule extends HelperRuntimeModule {
...
+					)});`
 				]),
 				"}",
-				modern
-					? "def['default'] = () => value;"
-					: "def['default'] = function() { return value; };",
+				`def['default'] = ${runtimeTemplate.returningFunction("", "value")};`,
 				`${RuntimeGlobals.definePropertyGetters}(ns, def);`,
 				"return ns;"
 			]),
```
Comment (alexander-akait) [inline lib/runtime/CreateFakeNamespaceObjectRuntimeModule.js:60]: Found more cases, refactored

**Comment by alexander-akait** [comment]
@sokra Done

**Comment by xinggangling** [comment]
> **What kind of change does this PR introduce?**
> 
> bugfix, but we still have `async` here, but not related to this fix, fixes #12885
> 
> **Did you add tests for your changes?**
> 
> hard to test, I think we can use `acorn` to check syntax (try to parse based on `es` target), if we will face with it again I will write couple tests
> 
> **Does this PR introduce a breaking change?**
> 
> No
> 
> **What needs to be documented once your changes are merged?**
> 
> No

`async` is also a problem; 
The ultimate goal of this issue is to avoid Es5 + syntax in the code。
IOS 9 WebView does not support

**Comment by xinggangling** [comment]
> **What kind of change does this PR introduce?**
> 
> bugfix, but we still have `async` here, but not related to this fix, fixes #12885
> 
> **Did you add tests for your changes?**
> 
> hard to test, I think we can use `acorn` to check syntax (try to parse based on `es` target), if we will face with it again I will write couple tests
> 
> **Does this PR introduce a breaking change?**
> 
> No
> 
> **What needs to be documented once your changes are merged?**
> 
> No

`async` is also a problem; 
The ultimate goal of this issue is to avoid Es5 + syntax in the code。
IOS 9 WebView does not support

**Comment by alexander-akait** [comment]
@xinggangling It is impossible, sorry, and out of scope PR, we have other issue for this

**Comment by xinggangling** [comment]
> @xinggangling It is impossible, sorry, and out of scope PR, we have other issue for this

topLevelAwait cause this problem,we had improved 
our code.😃

**Comment by alexander-akait** [comment]
@sokra I accidentally rewrite your changes using `force`, sorry :disappointed:

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Pull request merged**